src/__tests__/reducer.test.ts

```diff
@@ -0,0 +1,235 @@
+import { auth0Loaded, handleRedirectCallbackAction, loginWithPopupAction } from "../actions";
+import { auth0ProviderStateReducer } from "../reducer";
+import {
+  HandlingRedirectState,
+  isAuthenticatedState,
+  isLoadedState,
+  isUnauthenticatedState,
+  LoadingState,
+  TokenUser,
+  UnauthenticatedState,
+} from "../types";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const createFakeAuth0Client = (): any => {
+  return {
+    loginWithRedirect: jest.fn(),
+    getTokenWithPopup: jest.fn(),
+    getTokenSilently: jest.fn(),
+    getIdTokenClaims: jest.fn(),
+    logout: jest.fn(),
+  };
+};
+
+const createInitialState = (): LoadingState => ({
+  loading: true,
+  isAuthenticated: false,
+  popupOpen: false,
+  user: undefined,
+  auth0Client: undefined,
+  handlingRedirect: false,
+});
+
+const createFakeUnauthenticatedState = (): UnauthenticatedState => ({
+  isAuthenticated: false,
+  loading: false,
+  loginWithPopup: jest.fn(),
+  handleRedirectCallback: jest.fn(),
+  handlingRedirect: false,
+  user: undefined,
+  popupOpen: false,
+  // eslint-disable-next-line @typescript-eslint/camelcase,@typescript-eslint/ban-ts-ignore
+  // @ts-ignore
+  auth0Client: {
+    loginWithRedirect: jest.fn(),
+    getTokenWithPopup: jest.fn(),
+    getTokenSilently: jest.fn(),
+    getIdTokenClaims: jest.fn(),
+    logout: jest.fn(),
+  }, //new Auth0Client({ domain: "xyz", client_id: "pdq" }),
+});
+
+const createHandlingRedirectState = (): HandlingRedirectState =>
+  ({
+    ...createFakeUnauthenticatedState(),
+    handlingRedirect: true,
+  } as HandlingRedirectState);
+
+describe("reducer", () => {
+  describe("loginWithPopupAction", () => {
+    describe("started", () => {
+      it("flips the popupOpen flag to true", () => {
+        const origState = createFakeUnauthenticatedState();
+        const newState = auth0ProviderStateReducer(origState, loginWithPopupAction.started());
+        expect(newState.popupOpen).toBeTruthy();
+      });
+
+      it("throws an error if the auth0 state is not yet fully loaded", () => {
+        const origState = createInitialState();
+        expect(() =>
+          auth0ProviderStateReducer(origState, loginWithPopupAction.started()),
+        ).toThrowErrorMatchingSnapshot();
+      });
+    });
+
+    describe("done", () => {
+      it("flips the popupOpen flag to false", () => {
+        const origState = {
```

Comment:
with test-data-bot this could become

```ts
const originalState = createFakeUnauthenticatedState({
  overrides: { popupOpen: true },
})
```

Reply:
Nice improvement.  Maybe we can catch that one when we address #3